### PR TITLE
Groupnorm update to make better use of cores, and streamline the API

### PR DESCRIPTION
**Pull request by sadesoyeTT**
### Ticket
N/A

### Problem description.
Initial groupnorm implementation may not utilize the full available core grid. 
Frame of reference for core_grid was reversed.
DRAM Groupnorm API can use some improvements.

### What's changed

- Updated  DRAM group norm to maximize the available core grid and fix the frame of reference with core.x implying the columns (work done by @jonathansuTT)
- Added helper (`group_norm_params_from_torch`) function in the API to simplify the preparation affine weights and masks with support for sharding along a 1D mesh if desired. weight, bias and masks are prepped in one call.
Example: 
```
  [self.weight, self.bias], self.mask = ttnn.dram_group_norm_params_from_torch(
            torch_params=[state_dict["weight"], state_dict["bias"]],
            channels_per_device=self.num_channels,
            groups_per_device=self.num_groups,
            device=self.mesh_device,
            mesh_axis=self.mesh_axis,
            return_mask=True,
        )
```
- Added experimental heuristics for computing num_out_blocks in DRAM program code. Set num_out_blocks to -1

sdxl test on t3k:
```
2025-08-22 04:49:22.276 | INFO     | models.experimental.stable_diffusion_xl_base.tests.test_sdxl_accuracy:test_accuracy_sdxl:144 - Test results saved to test_reports/sdxl_test_results.json
PASSED
```
[Single device performance test](https://github.com/tenstorrent/tt-metal/actions/runs/17167203281)

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes

### CODE REVIEW

**Review by ipotkonjak-tt — COMMENTED** [review]
Could you please run [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) as this would change VAE device perf? And update the perf as well, to keep pipelines green.

The test is located here: `models/experimental/stable_diffusion_xl_base/tests/test_sdxl_perf.py`

The accuracy test does not check device perf

**Review by cglagovichTT — APPROVED** [review]

File: models/experimental/stable_diffusion_35_large/tests/fun_vae_decoder/test_fun_groupnorm.py [hunk 1]
```diff
@@ -94,7 +94,15 @@ def test_group_norm(
         sp_axis=sp_axis,
         tp_axis=tp_axis,
     )
-    device = parallel_manager.vae_parallel_config.device
+    vae_device = parallel_manager.submesh_devices[0]
+
+    if parallel_manager.dit_parallel_config.cfg_parallel.mesh_shape[1] != 4:
+        cfg_shape = parallel_manager.dit_parallel_config.cfg_parallel.mesh_shape
+        assert cfg_shape[0] * cfg_shape[1] == 4, f"Cannot reshape {cfg_shape} to a 1x4 mesh"
+        print(f"Reshaping submesh device 0 from {cfg_shape} to (1, 4) for CLIP + T5")
```
Comment (bbradelTT) [inline models/experimental/stable_diffusion_35_large/tests/fun_vae_decoder/test_fun_groupnorm.py]: Would you be able to use logger?

File: models/experimental/stable_diffusion_35_large/tt/fun_vae_decoder/fun_group_norm.py [hunk 2]
```diff
@@ -74,21 +74,28 @@ def from_torch(
         ), f"Incompatible channels ({num_channels}%32)or groups({num_channels}%{num_groups}):"
 
         opt_core_grid = parallel_config.device.core_grid
+        """
```
Comment (bbradelTT) [inline models/experimental/stable_diffusion_35_large/tt/fun_vae_decoder/fun_group_norm.py]: Should this code be removed?

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 3]
```diff
@@ -1214,55 +1214,72 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
...
+    // Compute optimal core grid
+    uint32_t num_virtual_cols = std::min<uint32_t>(grid_size.x, num_groups);
+    while ((W / num_virtual_cols) % TILE_WIDTH != 0) {
+        num_virtual_cols -= 1;
+    }
+    uint32_t num_actual_cols =
+        (grid_size.x / num_virtual_cols) * num_virtual_cols;  // Largest multiple of virtual cols < 8
+    uint32_t num_actual_rows = grid_size.y;
+    uint32_t num_virtual_rows =
+        (grid_size.x / num_virtual_cols) * num_actual_rows;  // TODO: Assert if we don't have enough data
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: Either fix this in this PR or provide an issue number here.

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 4]
```diff
@@ -1214,55 +1214,72 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
...
-    TT_FATAL(num_cores_c != 0, "num_cores_c should not equal 0");
-    TT_FATAL(num_cores_r != 0, "num_cores_r should not equal 0");
-    TT_FATAL(H % num_cores_r == 0, "width * height: {} must be divisible by num_cores.y: {}", H, num_cores_r);
-    TT_FATAL(W % num_cores_c == 0, "channels: {} must be divisible by num_cores.x: {}", W, num_cores_c);
+    uint32_t per_core_N = W / num_virtual_cols;
     uint32_t per_core_Mt_group_1 = per_core_M_group_1 / TILE_HEIGHT;
     uint32_t per_core_Mt_group_2 = 0;
     uint32_t per_core_Nt = (per_core_N + TILE_WIDTH - 1) / TILE_WIDTH;
-    uint32_t num_datum_row_per_group = W / num_groups;
+    uint32_t num_datum_row_per_group = W / num_groups;  // TODO: Rename to num_channels_per_group
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: Either fix this in this PR or provide an issue number here.

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 5]
```diff
@@ -1214,55 +1214,72 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
...
     uint32_t num_datum_row_per_group_mod_tile_w =
         num_datum_row_per_group % TILE_WIDTH == 0 ? TILE_WIDTH : num_datum_row_per_group % TILE_WIDTH;
-    uint32_t group_size = W / num_groups;
     // split each batch into multiple cores
-    uint32_t num_shards_r = H / per_core_M_group_1;
+    uint32_t num_shards_r = H / per_core_M_group_1;  // TODO: This should be num_virtual_rows
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: Either fix this in this PR or provide an issue number here.

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 6]
```diff
@@ -1214,55 +1214,72 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
...
     uint32_t num_cores_per_batch = num_batches > num_shards_r ? 1 : num_shards_r / num_batches;
-    uint32_t num_shards_c = W / per_core_N;
+    uint32_t num_shards_c = W / per_core_N;  // TODO: This should be num_virtual_cols
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: Either fix this in this PR or provide an issue number here.

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 7]
```diff
@@ -1450,15 +1463,15 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
 
     if (!equal_batches_per_core) {
         // input buffers
-        in0_block_tiles_group_1 = block_ht_group_1 / num_out_blocks * block_wt;
+        // in0_block_tiles_group_1 = block_ht_group_1 / num_out_blocks * block_wt;
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: remove?

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 8]
```diff
@@ -1214,55 +1214,72 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
...
     // each core contains multiple batches
     uint32_t num_batches_per_core_group_1 = num_batches > num_shards_r ? num_batches / num_shards_r : 1;
     uint32_t num_batches_per_core_group_2 = num_batches_per_core_group_1;  // need this to be non-zero even if unused
     uint32_t num_groups_per_core = num_groups > num_shards_c ? num_groups / num_shards_c : 1;
-    TT_FATAL(num_groups % num_cores_c == 0, "num_groups: {} must divide cores_y: {}", num_groups, num_cores_c);
+
+    // Compute num_out_blocks if not provided. If this does not provide the best result, num_out_blocks should be
+    // computed by testing different powers of 2 This is a heuristic.
+    if (num_out_blocks == -1) {
+        uint32_t den = (shape[1] * shape[2] * shape[3]) / (256 * 256 * (num_virtual_cols * num_virtual_rows));
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: Please have a well named const for 256.

**Review by bbradelTT — APPROVED** [review]

**Review by jonathansuTT — APPROVED** [review]

**Comment by sadesoyeTT** [comment]
> Could you please run [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) as this would change VAE device perf? And update the perf as well, to keep pipelines green.
> 
> The test is located here: `models/experimental/stable_diffusion_xl_base/tests/test_sdxl_perf.py`
> 
> The accuracy test does not check device perf

https://github.com/tenstorrent/tt-metal/actions/runs/17167203281. All green!
Your code will still end up using the total of the core grid that was passed in if 4x2, will use 8 for instance. Please change the core grid to 8x8 if SDXL has no other constraints. I didn't want to update the code without much context.

**Review by ipotkonjak-tt — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp:1]: @tt-ai please review changes in this file

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp:1]: **Claude finished @ayerofieiev-tt's task** —— [View job](https://github.com/tenstorrent/tt-metal/actions/runs/17216263870)

---
### Code Review for groupnorm_op_multi_core.cpp

#### Todo List:
- [x] Read and analyze the groupnorm_op_multi_core.cpp file
- [x] Review code changes for bugs, performance issues, and best practices
- [x] Check for issues mentioned in previous reviews (TODOs, magic numbers, unnecessary copies)
- [x] Provide comprehensive review feedback

#### Review Findings:

### 🔍 **Addressing Previous Review Comments**

**1. Clang-Tidy Issue - Unnecessary Copy (Line 2323):** ✅ **FIXED**
```cpp
const auto& virtual_group = mcast_virtual_groups[i];  // Now using const reference
```
Good fix! The variable is now properly declared as a const reference since it's only used for reading.

**2. Magic Number Issue (Lines 1265, 1270):** ⚠️ **NEEDS ATTENTION**
```cpp
const uint32_t heuristic_block_size_per_core = 256 * 256;  // Line 1265
while (num_out_blocks < heuristic_num_out_blocks && num_out_blocks < 256) {  // Line 1270
```
**Issue:** The magic number `256` appears multiple times without named constants.

**Recommendation:**
```cpp
static constexpr uint32_t MAX_OUT_BLOCKS = 256;
static constexpr uint32_t HEURISTIC_BLOCK_SIZE_BASE = 256;
const uint32_t heuristic_block_size_per_core = HEURISTIC_BLOCK_SIZE_BASE * HEURISTIC_BLOCK_SIZE_BASE;
```

**3. TODO Comment (Line 1221):** ⚠️ **NEEDS RESOLUTION** 
```cpp
(grid_size.x / num_virtual_cols) * num_actual_rows;  // TODO: Assert if we don't have enough data
```
This TODO should either be implemented or removed. Consider adding the assertion or creating a GitHub issue.

### 🚨 **Code Quality Issues**

**1. Core Grid Logic (Lines 1212-1221):**
The core grid computation logic has significant complexity that would benefit from validation:

```cpp
uint32_t num_virtual_cols = std::min<uint32_t>(grid_size.x, num_groups);
while ((W / num_virtual_cols) % TILE_WIDTH != 0) {
    num_virtual_cols -= 1;
}
uint32_t num_actual_cols = (grid_size.x / num_virtual_cols) * num_virtual_cols;
uint32_t num_virtual_rows = (grid_size.x / num_virtual_cols) * num_actual_rows;
```

**Concerns:**
- The `while` loop could potentially reach 0 if no valid `num_virtual_cols` exists
- `num_virtual_rows` calculation is complex and could benefit from clearer variable naming
- Missing validation for edge cases

**2. Potential Copy Issues (Line 2322):**
```cpp
auto group = mcast_groups[i];  // This creates a copy of the vector
```
**Issue:** This creates an unnecessary copy of the vector. Should be:
```cpp
const auto& group = mcast_groups[i];
```

**3. Variable Shadowing (Lines 2327-2328):**
```cpp
CoreCoord core = group[j];
CoreCoord virtual_core = virtual_group[j];
```
These could also be const references if they're not modified:
```cpp
const CoreCoord& core = group[j];
const CoreCoord& virtual_core = virtual_group[j];
```

### ✅ **Positive Changes**

**1. Improved Core Utilization:**
The updated logic for `num_virtual_rows` and `num_virtual_cols` appears to better utilize the available core grid as intended.

**2. Frame of Reference Fix:**
The coordinate system appears to have been corrected with proper x/y mapping for cores.

**3. Heuristic Block Computation:**
The experimental heuristics for `num_out_blocks` (lines 1264-1273) provide a reasonable fallback when `-1` is passed.

### 📋 **Recommendations Summary**

1. **HIGH PRIORITY:**
   - Replace magic numbers with named constants
   - Fix unnecessary copy in line 2322: `const auto& group = mcast_groups[i];`
   - Resolve or remove TODO comment (line 1221)

2. **MEDIUM PRIORITY:**
   - Add validation for the `while` loop in core grid computation
   - Consider making `core` and `virtual_core` const references if not modified
   - Add more descriptive variable names for complex calculations

3. **LOW PRIORITY:**
   - Consider extracting the core grid computation logic into a separate function for better testability
   - Add more detailed comments explaining the heuristic calculations

### 💡 **Additional Notes**

The overall changes align with the PR's goals of better core utilization and API streamlining. The code is functionally sound, but the identified issues should be addressed for better maintainability and code quality.

---

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 9]
```diff
@@ -1202,32 +1202,53 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
...
+        num_virtual_cols,
+        W,
+        num_groups,
+        TILE_WIDTH);
+    TT_FATAL(
+        H >= num_virtual_rows,
+        "Total size of a slice across channel dimension:{} must be greater than or equal to num_virtual_rows: {}. "
+        "Reduce grid_size as needed",
+        H,
+        num_virtual_rows);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: These seem to be catching a programming error, not user error.

One of important parts of errors - user should know what to do about the error.
See this guide:
https://github.com/tenstorrent/tt-metal/blob/main/contributing/ErrorMessageBestPractices.md

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 10]
```diff
@@ -1202,32 +1202,53 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
...
     const auto& shape = a.padded_shape();
-    uint32_t H = shape[2] * num_batches;
+    uint32_t H = shape[1] * shape[2] * num_batches;
     uint32_t Ht = H / TILE_HEIGHT;
     uint32_t W = shape[3];
     uint32_t Wt = W / TILE_WIDTH;
-    uint32_t per_core_M_group_1 = H / num_cores_r;
+    // Compute optimal core grid
+    uint32_t num_virtual_cols = std::min<uint32_t>(grid_size.x, num_groups);
+    while ((W / num_virtual_cols) % TILE_WIDTH != 0) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: Can you help to understand if it is impossible to hit a division by 0 here?

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: might want smth like this instead

```cpp
uint32_t compute_num_virtual_cols(uint32_t W, uint32_t TILE_WIDTH, uint32_t grid_x, uint32_t num_groups) {
    uint32_t limit = std::min<uint32_t>(grid_x, num_groups);
    if (limit == 0 || TILE_WIDTH == 0) return 0;
    if (W == 0) return limit;                 // 0 is divisible by anything
    if (W % TILE_WIDTH != 0) return 0;        // impossible to satisfy

    uint32_t base = W / TILE_WIDTH;           // need n | base and n <= limit
    limit = std::min(limit, base);

    uint32_t best = 0;
    for (uint32_t d = 1; d <= base / d; ++d) {
        if (base % d == 0) {
            uint32_t d1 = d, d2 = base / d;   // pair of divisors
            if (d1 <= limit && d1 > best) best = d1;
            if (d2 <= limit && d2 > best) best = d2;
        }
    }
    return best;  // 0 if none (shouldn’t happen if W%TILE_WIDTH==0)
}
```

**Review by vsureshTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp (hunk 10 above)
Comment (sadesoyeTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: I have moved the asserts around to prevent it from happening.
